internal/config/main.go

```diff
@@ -16,28 +16,68 @@ package config
 
 import (
 	"fmt"
+	"time"
 
-	"github.com/open-telemetry/opentelemetry-operator/internal/version"
+	"github.com/go-logr/logr"
 	"github.com/spf13/pflag"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+
+	"github.com/open-telemetry/opentelemetry-operator/internal/version"
+	"github.com/open-telemetry/opentelemetry-operator/pkg/autodetect"
+	"github.com/open-telemetry/opentelemetry-operator/pkg/platform"
+)
+
+const (
+	defaultAutoDetectFrequency     = 5 * time.Second
+	defaultCollectorConfigMapEntry = "collector.yaml"
 )
 
 // Config holds the static configuration for this operator
 type Config struct {
-	v                       version.Version
+	// Registers a callback, to be called once a configuration change happens
+	OnChange func() error
+
+	logger              logr.Logger
+	autoDetect          autodetect.AutoDetect
+	autoDetectFrequency time.Duration
+	onChange            []func() error
+
+	// config state
 	collectorImage          string
 	collectorConfigMapEntry string
+	plt                     platform.Platform
+	v                       version.Version
 }
 
-// DefaultConfig builds a new configuration populated with the default, immutable values
-func DefaultConfig() Config {
-	return WithVersion(version.Get())
-}
+// New constructs a new configuration based on the given options
+func New(opts ...Option) Config {
+	// initialize with the default values
+	o := options{
+		autoDetectFrequency:     defaultAutoDetectFrequency,
+		collectorConfigMapEntry: defaultCollectorConfigMapEntry,
+		logger:                  logf.Log.WithName("config"),
+		plt:                     platform.Unknown,
+		v:                       version.Get(),
```

Comment:
I was unsure whether to keep using `v` or `version`. The downside of using `version` is that it shadows the package name `version`, but I could actually rename the import to `v` in case it needs to be used after the shadowing:

```go
version := version.Get()
another_version := version.Get() // this isn't possible anymore, as 'version' (the package) is now shadowed by the var
```

Given that this is a field in the object, this situation shouldn't be frequent anyway.

Reply:
Done!